### PR TITLE
Introduce a config builder for OIDC Client Registration

### PR DESCRIPTION
**Pull request by michalvavrik**
This is but part of phase 2 agreed here https://github.com/quarkusio/quarkus/issues/39185#issuecomment-2439669375.
We agreed to start with the OIDC Client Registration because its API is experimental, therefore we can make changes on it and it's better to make breaking changes sooner than later.
This PR:

- introduces builder for `OidcClientRegistrationConfig` config group that users will use when they register new OIDC Client
- migrates `io.quarkus.oidc.client.registration.OidcClientRegistrations` from former config group class to config group interface
- `OidcCommonConfig` now implements `@ConfigMapping` config group interface, which means we can use the interface without converting it to the class when not necessary (and we are not breaking anything); this needs to be done so that OIDC Common utils can accept both former class and new config

When this PR is merged, it will be easier to follow with similar changes for the OIDC and OIDC Client extensions because they share `OidcCommonConfig`.

### CODE REVIEW

**Comment by sberyozkin** [comment]
Hi @michalvavrik Looks good, thanks, how will a more complex sequence look like, where some metadata are supplied as well ?

**Comment by michalvavrik** [comment]
> Hi @michalvavrik Looks good, thanks, how will a more complex sequence look like, where some metadata are supplied as well ?

It's flat structure because metadata has so little properties that I thought it was overkill to create extra `.metadata().clientName("name").redirect....`.

You could find some examples in tests that I added, but basically this is how it works:

```
var config = OidcClientRegistrationConfig.builder()
                .extraProperty("metadata-extra-prop-key-1", "metadata-extra-prop-val-1")
                // or
                .extraProps(Map.of("one", "one", "two", "two"))
                .clientName("metadata-client-name")
                .redirectUri("metadata-redirect-uri")
                .postLogoutUri("metadata-post-logout-uri")
                // and now continue with wehatever you like
                .id("client-id-1")
                .build();
```

**Comment by sberyozkin** [comment]
@michalvavrik See, there will be a lot of such sub-interfaces for `OidcTenantConfig`, so now we have an opportunity to work out now how to make it as easy as possible for the users. I'm not too keen on having everything flat, IMHO, `metadata` should have its own build support to be distinctive from the rest of the key properties like registration url, and what would be nice is to be able to do either `metadata().clientName()...` as part of the main builder sequence, or for the main builder sequence to do `.metadata(metadata)` where `metadata` was built independently.

I'd also like to suggest to offer shortcuts to let users minimize a number of `.builder()`.
For example, `OidcClientRegistrationConfig.registrationUri(someUrl)` returns a Builder...

**Comment by michalvavrik** [comment]
> @michalvavrik See, there will be a lot of such sub-interfaces for `OidcTenantConfig`, so now we have an opportunity to work out now how to make it as easy as possible for the users. I'm not too keen on having everything flat, IMHO, `metadata` should have its own build support to be distinctive from the rest of the key properties like registration url, and what would be nice is to be able to do either `metadata().clientName()...` as part of the main builder sequence, or for the main builder sequence to do `.metadata(metadata)` where `metadata` was built independently.
> 
> I'd also like to suggest to offer shortcuts to let users minimize a number of `.builder()`. For example, `OidcClientRegistrationConfig.registrationUri(someUrl)` returns a Builder...

Alright, I'll look into it.

**Comment by michalvavrik** [comment]
Hello @sberyozkin , I believe I have addressed your feedback.

File: extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilder.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        this.initialToken = Optional.ofNullable(initialToken);
+        return this;
+    }
+
+    /**
+     * @param clientName {@link OidcClientRegistrationConfig.Metadata#clientName()}
+     * @param redirectUri {@link OidcClientRegistrationConfig.Metadata#redirectUri()}
+     * @return this builder
+     */
+    public OidcClientRegistrationConfigBuilder metadata(String clientName, String redirectUri) {
```
Comment (sberyozkin) [inline extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilder.java:124]: Hi @michalvavrik, We now have a dedicated `Metadata` builder which is good, and this is a good shortcut, I propose to have one more `metadata(String redirectUri)` and that should... and we can follow the same pattern for `OidcTenantConfig`...

There is something I'd like to explore....

I'd like to type something like the following, which might be not too bad for bigger config mappings, for example:
```
 OidcClientRegistrationConfig clientRegConfig = OidcClientRegistrationConfig
                    .authServerUrl(authServerUrl)
                    .metadata()
                        .name("My Client")
                        .redirectUri("https://localhostL8443/redirect")
                        .build();
                    .build();
``` 

I.e, we introduce a convention that a no arg sub-ConfigMapping like `metadata()` creates a builder which returns a variation of `MetadataBuilder`, which, once its `build()` is called will return `OidcClientRegistrationConfigBuilder`, for example, let's call it `OidcClientRegistrationConfigMetadataBuilder`...

So users will have an option to build the whole instance in one structured sequence, for example,
```
OidcTenantConfig build() {
    return 
    OidcTenantConfig.authServerUrl("someurl")
        .tenanId("mytenant")
        .authentication()
            .a("aaa")
            .b("bbb")
            .build)
        .token()
            .a("aaa")
            .b("bbb")
            .build()
     .build();
}
```

When users have to deal with many sub ConnfigMappings and each ConfigMapping can have 5-10 properties, the above can save users some space in their code and be quite optimal...

Comment (michalvavrik) [reply extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilder.java:124]: Hey @sberyozkin , I'll do it myself. I just hope I get you right. I am starting on it right now, but for context, this is how I understand requests.

1. add `metadata(String redirectUri)` - I am on it
2. `I'd like to type something like the following, which might be not too bad for bigger config mappings, for example:` but you can do that, I am not certain if I understand you. Please have a look here and point out based on that example what you would like to change. I'll do it.

- https://github.com/quarkusio/quarkus/blob/09aa6c30aa5eb4399fcce36c3ecb62dcd55593f0/extensions/oidc-client-registration/runtime/src/test/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilderTest.java#L71
- https://github.com/quarkusio/quarkus/blob/09aa6c30aa5eb4399fcce36c3ecb62dcd55593f0/extensions/oidc-client-registration/runtime/src/test/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilderTest.java#L161
- https://github.com/quarkusio/quarkus/blob/09aa6c30aa5eb4399fcce36c3ecb62dcd55593f0/extensions/oidc-client-registration/runtime/src/test/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilderTest.java#L183
- https://github.com/quarkusio/quarkus/blob/09aa6c30aa5eb4399fcce36c3ecb62dcd55593f0/extensions/oidc-client-registration/runtime/src/test/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilderTest.java#L256
3. `So users will have an option to build the whole instance in one structured sequence, for example,` I still think that is what is done here. Only difference is that I can see is that I differentiate between `end` and `build`. But how else to do it?

Comment (michalvavrik) [reply extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilder.java:132]: Done, I have added `metadata(String redirectUri)` and test for it.

Comment (sberyozkin) [reply extensions/oidc-client-registration/runtime/src/main/java/io/quarkus/oidc/client/registration/OidcClientRegistrationConfigBuilder.java:132]: @michalvavrik Oh it is great, I missed `.end()` for a metadata() sub-builder...

**Review by sberyozkin — APPROVED** [review]

**Pull request merged**